webapp/src/app/participant-view/participant-stage-view/exp-chat/exp-chat.component.ts

```diff
@@ -72,12 +73,51 @@ const TIMER_SECONDS = 60; // 1 minute between item pairs for discussions
   styleUrl: './exp-chat.component.scss',
 })
 export class ExpChatComponent implements OnDestroy {
+  // Reload the internal logic dynamically when the stage changes
+  @Input({ required: true })
+  set stage(value: ExpStageChatAboutItems) {
+    this._stage = value;
+
+    this.everyoneReachedTheChat = this.participant.everyoneReachedCurrentStage(this.stage.name);
+
+    // Initialize the current rating to discuss with the first available pair
+    const { id1, id2 } = this.stage.config.ratingsToDiscuss[0];
+    this.currentRatingsToDiscuss = signal({
+      item1: this.stage.config.items[id1],
+      item2: this.stage.config.items[id2],
+    });
+
+    this.unsubscribeMessages = chatMessagesSubscription(this.stage.config.chatId, (m) => {
+      this.messages.set(extendUntilMatch(this.messages(), m.reverse(), 'uid'));
```

Comment:
We should maybe chat about using `extendUntilMatch`; from looking at the code, I think it's not strictly an extension; it seems to take old elements up to the when the 'key' (in this case 'uid' in new list matched the old list, and then concat from that point on)... this makes me think this logic is probably slightly wrong as I think two messages by the same person will result in the cut being too early... but I'm not sure if I'm getting the logic... 

I guess my expectation here is that new messages just replace the old ones... why do we need/benefit from this fancy concat logic?

Maybe good to comment on what "m" is, is it a single message? or a set of messages? Maybe a longer variable name would help?

Reply:
currently, when a participant opens a chat stage, they subscribe to the `n` most recent messages from firestore. On every update, the frontend appends new messages to its list (because from my understanding of firestore, the situation where 2 messages being created in a short span could only trigger one subscription event can happen ?). We can indeed discuss of better ways to handle incoming messages

Reply:
`extendUntilMatch` assumes that 2 arrays of objects are "sorted" in a compatible order, and takes from the first array until an object with the same key value as the second array's first element is reached. It then appends all of the second array. It does not do so "in-place", so I agree that "extend" may not be the right word for it...

Reply:
But isn't "uid" a user's ID? or is it a unique firestore object ID? I think if you did this with a unique message ID, it might be right... But I would suggest a simpler way to do this: take the messages, push into a map by message ID, and then Object.assign to that to get the union, and then re-flatten by message sent timestamp? Then you don't depend on subtle semantics of extendUntilMatch... An important part of this kind of coding is making the semantics as easy to interpret as possible (more time is generally spent reading code than writing code, so optimise for each of reading) ) :)

Reply:
in this code, `uid` stands for `unique ID`, I agree that it is misleading. Perhaps using `mid` for `message ID` would be better ?

Reply:
Ah, we may not control that if it's a firebase value... but at least a comment to clarify since uid is so often used for user-id also.

Reply:
I could define this function and use it instead of `extendUntilMatch`
```typescript
export const mergeByKey = <T extends object, K extends Indexable<T>>(
  previousArray: T[],
  incomingArray: T[],
  key: K,
): T[] => {
  return Object.values({ ...lookupTable(previousArray, key), ...lookupTable(incomingArray, key) });
};
```

Reply:
done in fbd6f6420695b44e45e2662ed8927502f859e681